Show Loader when adding Workspace Edit

mananjadhav opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
- Add loading flags for Avatar upload and form submission
- IMP: Added `isUploading` flag to AvatarWithImagePicker with uploadIndicator
- Replaced js loop to Animated.loop

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/4723
--->
$ https://github.com/Expensify/App/issues/4487

### Tests
1. Tested by keeping the form contents the same
2. Updating the workspace name and image
3. Removing the workspace image
4. Submitting the form successfully
5. Uploading a large image

<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### QA Steps
1. Go to Workspace and click on the Avatar
2. Check button state when the values are not updated
3. Change the content/upload the image and see if values are updated
3. Test Removing the workspace image as well 
4. Submitting the form successfully
5. Uploading a large image

<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
**v1**
https://user-images.githubusercontent.com/3069065/131385161-e70fc702-9ef7-4515-bacd-86ac728d5c03.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132609914-524caa8b-ead7-4ec1-97c4-6288d6575dd7.mp4

#### Mobile Web
**v1**
https://user-images.githubusercontent.com/3069065/131384341-a5d4da09-e5e7-4807-8cf9-3724c797d2c6.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610024-500c89e5-83d0-4c05-91c6-61688fd223c1.mp4

#### Desktop
**v1**
https://user-images.githubusercontent.com/3069065/131384466-c488f178-844d-471b-b9b6-9ca68fa5de65.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610083-e81b6468-c976-4ec2-83cc-f7932de947f3.mp4

#### iOS
**v1**
https://user-images.githubusercontent.com/3069065/131384510-c38701d8-74ae-49c2-a116-e50e2fdc19a5.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610283-1a851e62-640f-496d-86d2-83376a6e7125.mp4

#### Android
**v1**
https://user-images.githubusercontent.com/3069065/131384679-c6b796b0-f49e-4e76-8d48-3991d2cac241.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610438-29e4b86c-81aa-45fc-ad41-7d77e0ba6c92.mp4

mananjadhav committed:
fix(wkspace-edit-loader): Added loader for AvatarWithImagePicker

mananjadhav committed:
fix(wkspace-edit-loader): Added loader for avatar + form submission

mananjadhav commented:
@roryabraham I've raised a draft PR. Need to clarify a couple of things:

1. Should we throw error messages on any form error or avatar upload? or even a success message?
2. I've noticed when we upload an Avatar image, there isn't any feedback shown to the user. For now, what I've done is added an `isUploading` flag to `AvatarWithImagePicker` which shows the `Sync` icon instead of Pencil when set true. While it did solve the purpose, it doesn't look great. I would recommend an animation similar to `AvatartWithIndicator`. Shall I go ahead and add it?
3. If we agree to do the loader thing in #2, a similar UX is missing for the Profile image update as well. Shall I cover that too in this PR if is a 🟢 for the above approach.

roryabraham commented:
> Should we throw error messages on any form error or avatar upload? or even a success message

If there's a form error or API error, we should display a user-friendly error message. A success message would be good too, maybe using [Growl](https://github.com/Expensify/App/blob/2bde5c1ce2c6563676e47ff061f4dd4861bf1a19/src/libs/Growl.js) or some other visual indication.

> I've noticed when we upload an Avatar image, there isn't any feedback shown to the user. For now, what I've done is added an isUploading flag to AvatarWithImagePicker which shows the Sync icon instead of Pencil when set true. While it did solve the purpose, it doesn't look great. I would recommend an animation similar to AvatartWithIndicator. Shall I go ahead and add it?

I'm not sure. I'll defer to @Expensify/design here

> If we agree to do the loader thing in Fix spaces #2, a similar UX is missing for the Profile image update as well. Shall I cover that too in this PR if is a 🟢 for the above approach?

I assume you meant item 2 of your list, not the PR linked in the comment. I would think yes 🟢, but again will defer to @Expensify/design here.

---

Lastly, I'll warn you that what I see in this PR so far violates the [Expensify/App Philosophy](https://github.com/Expensify/App#philosophy), specifically `UI Binds to Disk`. My interpretation of that philosophy is that actions should never really be returning a promise – instead the UI binds to data in Onyx, and calls actions that manage the data. It should never be explicitly waiting for a response.

mananjadhav commented:
Thanks for the clarification @roryabraham.


 
> If there's a form error or API error, we should display a user-friendly error message. A success message would be good too, maybe using [Growl](https://github.com/Expensify/App/blob/2bde5c1ce2c6563676e47ff061f4dd4861bf1a19/src/libs/Growl.js) or some other visual indication.

I'll put in a message.

> I'm not sure. I'll defer to @Expensify/design here

 I'll wait for the comments from the design team.

> I assume you meant item 2 of your list, not the PR linked in the comment. 

Yes. I meant point# 2 from my list. It seems GH linked to the PR due to #.

> Lastly, I'll warn you that what I see in this PR so far violates the [Expensify/App Philosophy](https://github.com/Expensify/App#philosophy), specifically `UI Binds to Disk`. 

Thanks for highlighting.  I only went ahead with the approach because the same file had the following line:

```
this.uploadAvatarPromise = uploadAvatar(image).then(url => new Promise((resolve) => {
```
So should I store these loading props to Onyx?

roryabraham commented:
> So should I store these loading props to Onyx?

For now, I believe this is the approach we'll want to take.

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into fix/wkspace-edit-loader

mananjadhav committed:
fix(wkspace-edit-loader): Added upload indicator to AvatarWithImagePicker

mananjadhav committed:
fix(wkspace-edit-loader):  Added loading indicator to Profile Image

mananjadhav committed:
fix(wkspace-edit-loader): Changed flags from state to onyx props

mananjadhav committed:
fix(wskpace-edit-loader): Added function for local updates of policy values

mananjadhav committed:
fix(wkspace-edit-loader): Removed return Promise and added catch to show error

mananjadhav committed:
fix(wkspace-edit-loader): Added catch to throw error

mananjadhav commented:
@roryabraham I've updated the PR with the following changes:
1. Growl messages in catch block
2. Moving all flags to Onyx
3. Uploading Indicator in the AvatarImagePicker


I haven't added Growl on success because immediately after the policy is saved, we redirect to workspace card with updated info. Let me know if you still feel we should add a Growl message for success.

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into fix/wkspace-edit-loader

roryabraham commented:
@mananjadhav No growl on success sounds fine to me 👍

stitesExpensify commented on src/pages/workspace/WorkspaceEditorPage.js:
```diff
@@ -50,26 +51,35 @@ class WorkspaceEditorPage extends React.Component {
 
         // Store the upload avatar promise so we can wait for it to finish before updating the policy
         this.uploadAvatarPromise = uploadAvatar(image).then(url => new Promise((resolve) => {
+            updateLocalPolicyValues(this.props.policy.id, {isAvatarUploading: false});
```
We can prevent repeating this line by putting in in an `always()` right?

mananjadhav replied:
Can you elaborate this please?

mananjadhav replied:
Got it. Sure. Will update.

mananjadhav replied:
Done

stitesExpensify commented on src/pages/workspace/WorkspaceEditorPage.js:
```diff
@@ -50,26 +51,35 @@ class WorkspaceEditorPage extends React.Component {
 
             update(policyID, {name, avatarURL});
+        }).catch(() => {
+            updateLocalPolicyValues(this.props.policy.id, {isPolicyUpdating: false});
```
Since there isn't one of these in the `then()` how does the `isPolicyUpdating` get changed when the promise does not fail?

mananjadhav replied:
It's in the `then` block of `update` function in the API call.

stitesExpensify replied:
Gotcha 👍

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into fix/wkspace-edit-loader

mananjadhav committed:
fix(wkspace-editor-loader): Fixed icon size for loader

mananjadhav committed:
fix(wkspace-edit-loader): Fixed icon sizes in indicator

mananjadhav commented:
@stitesExpensify @roryabraham I've updated the PR. Additionally, I've also modified the size of the indicator to get it closed to the Avatar and adjusted the icon size.


https://user-images.githubusercontent.com/3069065/131907746-2cb1af94-8dd4-45c3-a338-ddc49f27863e.mov

stitesExpensify reviewed (dismissed):
LGTM! All you @roryabraham

roryabraham commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -51,17 +61,97 @@ const defaultProps = {
+     */
+    startRotation() {
+        this.rotate.setValue(0);
+        Animated.timing(this.rotate, {
```
Rather than using recursion here, it's better to use `Animated.loop` because it won't block the JS thread to make the recursive call:

```javascript
Animated.loop(
    Animated.timing(this.rotate, {
        toValue: 1,
        duration: 2000,
        easing: Easing.linear,
        isInteraction: false,
        useNativeDriver: true,
    }),
).start();
```

roryabraham replied:
i.e: the animation will restart immediately in the UI thread, instead of:

1. Sending a message through the bridge back to JS to say the animation is done
1. JS thread is blocked to execute the callback
1. Callback sends a message back through the bridge to UI thread to restart the animation.

mananjadhav replied:
Thanks for the clarification. I haven't worked with Animation earlier. It seems though we're using the same thing in `AvatarWithIndicator`. What should we do about that?

mananjadhav replied:
Updated and works great. Thanks for the suggestion. 🙌

roryabraham commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -115,26 +216,45 @@ class AvatarWithImagePicker extends React.Component {
-                                    animationIn="fadeInDown"
-                                    animationOut="fadeOutUp"
-                                />
+                                {
```
Spacing is just a little weird here (should be 4 spaces past the `<>` not 2):

```suggestion
                                  {
```

mananjadhav replied:
Will check. Not sure how editor didn't pick this.

mananjadhav replied:
@roryabraham It seems the `eslint` rolls back to two spaces when we use `<>`.  I also checked across the app, it seems that we're using two spaces only after `<>`. 

![image](https://user-images.githubusercontent.com/3069065/132085113-b5607d19-947e-4f61-a6d9-179cb13be230.png)
![image](https://user-images.githubusercontent.com/3069065/132085121-0064d234-e87d-4474-9f5e-fa34ad40900a.png)
![image](https://user-images.githubusercontent.com/3069065/132085126-1423fc8f-33f0-49f5-9e7c-96c379147877.png)

Please let me know if you meant anything else.

roryabraham replied:
Okay, no worries about the spacing then, we can just trust ESLint.

roryabraham commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -115,26 +216,45 @@ class AvatarWithImagePicker extends React.Component {
-                                    animationOut="fadeOutUp"
-                                />
+                                {
+                              this.props.isUploading
```
Also, this should probably be indented 4 spaces past the enclosing `{}`:

Like this:

```javascript
{
    this.props.isUploading
        ? ()
        : ()
}
```

mananjadhav replied:
Fixed this

roryabraham reviewed (changes requested):
Just had a few small comments, overall this looks great!

mananjadhav committed:
fix(wkspace-edit-loader): Changed recursive func call to Animate.loop

mananjadhav commented:
PR updated

roryabraham reviewed (dismissed):
Thanks @mananjadhav, looks good. If you wanted to update AvatarWithIndicator to use Animated.loop for consistency and a small performance boost that would be great!

Someday I hope we can switch over to using Reanimated instead of React Native's Animated. It comes with much better animation performance and in my opinion a simpler API. But that's a conversation for another day 😄

mananjadhav commented:
I’ll update the AvatarWithIndicator too. 

Yeah I’ve worked on Reanimated v1. But I’ve heard Reanimated V2 last year reduces a lot of boilerplate code too.

mananjadhav committed:
fix(wkspace-edit-loader): Changed recursive func call to Animate.loop in AvatarIndicator

mananjadhav commented:
PR updated

roryabraham reviewed (changes requested):
Okay, I apologize for not noticing these things earlier, but I have a few more comments:

### Start both animations in parallel
I just noticed that there are two animations that are basically happening in parallel, so we should just use `Animated.parallel`, something like this:

```javascript
/**
 * Start Animation for Indicator
 *
 * @memberof AvatarWithImagePicker
 */
start() {
    Animated.loop(
        Animated.parallel([
            Animated.timing(this.rotate, {
                toValue: 1,
                duration: 2000,
                easing: Easing.linear,
                isInteraction: false,
                useNativeDriver: true,
            }),
            Animated.spring(this.scale, {
                toValue: 1.666,
                tension: 1,
                isInteraction: false,
                useNativeDriver: true,
            }),
        ]),
    ).start();
}

/**
 * Stop Animation for Indicator
 *
 * @memberof AvatarWithImagePicker
 */
stop() {
    Animated.spring(this.scale, {
        toValue: 1,
        tension: 1,
        isInteraction: false,
        useNativeDriver: true,
    }).start(() => {
        this.rotate.resetAnimation();
        this.scale.resetAnimation();
        this.rotate.setValue(0);
    });
}
```

### DRY up animation code

To do this, I think we should:

1. Create a new file at `src/styles/animations/SpinningIndicatorAnimation.js`
1. That file should export a single class called `SpinningIndicatorAnimation`:

    ```javascript
      import {Animated, Easing} from 'react-native';

	  class SpinningIndicatorAnimation {
	      constructor() {
	          this.rotate = new Animated.Value(0);
	          this.scale = new Animated.Value(1);
	          this.start = this.start.bind(this);
	          this.stop = this.stop.bind(this);
	          this.getSyncingStyles = this.getSyncingStyles.bind(this);
	      }
	  
	      /**
	       * Start Animation for Indicator
	       *
	       * @memberof AvatarWithImagePicker
	       */
	      start() {
	          Animated.loop(
	              Animated.parallel([
	                  Animated.timing(this.rotate, {
	                      toValue: 1,
	                      duration: 2000,
	                      easing: Easing.linear,
	                      isInteraction: false,
	                      useNativeDriver: true,
	                  }),
	                  Animated.spring(this.scale, {
	                      toValue: 1.666,
	                      tension: 1,
	                      isInteraction: false,
	                      useNativeDriver: true,
	                  }),
	              ]),
	          ).start();
	      }
	  
	      /**
	       * Stop Animation for Indicator
	       *
	       * @memberof AvatarWithImagePicker
	       */
	      stop() {
	          Animated.spring(this.scale, {
	              toValue: 1,
	              tension: 1,
	              isInteraction: false,
	              useNativeDriver: true,
	          }).start(() => {
	              this.rotate.resetAnimation();
	              this.scale.resetAnimation();
	              this.rotate.setValue(0);
	          });
	      }
	  
	      /**
	       * Get Indicator Styles while animating
	       *
	       * @returns {Object}
	       */
	      getSyncingStyles() {
	          return {
	              transform: [{
	                  rotate: this.rotate.interpolate({
	                      inputRange: [0, 1],
	                      outputRange: ['0deg', '-360deg'],
	                  }),
	              },
	              {
	                  scale: this.scale,
	              }],
	          };
	      }
	  }
	  
	  export default SpinningIndicatorAnimation;
    ```
    
1. Then both `AvatarWithImagePicker` and `AvatarWithIndicator` can just do:

    ```javascript
    constructor(props) {
        super(props);
        this.animation = new SpinningIndicatorAnimation();
        ...
        ...
    }
    ```
    
    And then replace all uses of the animation methods with just `this.animation.start()`, `this.animation.stop()`, or `this.animation.getSyncingStyles`.

### Scaling issue
Lastly, I'm looking at the screenshots provided in the PR description, and it doesn't seem like the indicator is getting bigger when the animation starts. Instead, it actually looks like it's getting smaller and is outside the avatar circle.

mananjadhav commented:
Okay, I'll check the Animation updates and implement them.

> Lastly, I'm looking at the screenshots provided in the PR description, and it doesn't seem like the indicator is getting bigger > when the animation starts. Instead, it actually looks like it's getting smaller and is outside the avatar circle.

I had updated this comment but not the PR description. Let me update it in the PR description.

https://github.com/Expensify/App/pull/4801#issuecomment-912008496

mananjadhav commented:
@roryabraham There's a comment in `AvatarWithIndicator.startRotation`,

```
We need to manually loop the animations as `useNativeDriver` does not work well with `Animated.loop`.
```

mananjadhav committed:
fix(wkspace-edit-loader): Moved animation to separate class with Animation.parallel

mananjadhav committed:
fix(wkspace-edit-loader): Replace animation with SpinningAnimation

mananjadhav commented:
@roryabraham I am updating the `SpinnerAnimation` and having the following issue with `AvatarWithIndicator`, which I can't reproduce in staging.new.expensify.com

If I add a couple of messages in Offline mode, and then disable throttling, the `Avatar` scale doesn't change back to 1. (stays 1.66 and hence it looks bigger), neither does the rotation stop. I've pushed my change and attaching the video of the issue.


https://user-images.githubusercontent.com/3069065/132413898-d7d3abb1-168b-4c01-af82-75e828a6f056.mov


I am guessing it is because after `this.animation.stop()` the component isn't rendered and `getIndicatorStyles` don't have the updated values of `this.scale` and `this.rotate`.  I also feel `Animated.parallel` use must be causing it. I am a little confused that I am unable to reproduce the same in the staging URL where we have the old Animation in place.

mananjadhav committed:
fix(wkspace-edit-loader): Rollback Animated.parallel and Animated.loop changes

mananjadhav commented:
@roryabraham Did you get a chance to look at my comments above?

Meanwhile I'll resolve the conflicts.

roryabraham commented:
@mananjadhav Regarding this comment:

> We need to manually loop the animations as `useNativeDriver` does not work well with `Animated.loop`.

is there some observable problem that occurs when you use `Animated.loop` as I suggested above? I looked back at the PR which added that comment and there wasn't any additional context explaining the problem. AFAICT there's nothing that seems to go wrong when I switch to using `Animated.loop` locally, and it makes the code clearer and more performant. So I'd prefer to do that unless we can somehow expand on how "`useNativeDriver` does not work well with `Animated.loop`". Hopefully that makes sense!

We can skip the use of `Animated.parallel` though since you've found an issue with my suggested implementation.

mananjadhav commented:
Well, I didn't find any issue on the emulator but I haven't been able to check on a real Android device. Let me see if I can test and accordingly update the PR.

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into fix/wkspace-edit-loader

# Conflicts:
#	src/libs/actions/PersonalDetails.js

mananjadhav committed:
fix(wkspace-edit-loader):  Changed to Animated.loop

mananjadhav commented:
@roryabraham PR updated. I checked on a real Android device and it worked fine. So I am guessing it shouldn't be a problem.

I've also updated the videos with the latest testing of indicator closer to the image with increased size.

roryabraham reviewed (approved):
Okay, thanks for making all those changes, this is looking great!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1218882715) to staging by @roryabraham in version: 1.0.95-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

botify commented:
This has been deployed to production and is now subject to a 7-day regression period.
If no regressions arise, payment will be issued on 2021-09-17.  :confetti_ball:

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1222459337) to production by @roryabraham in version: 1.0.96-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
